src/kms/DeterministicCrypto.ts

```diff
@@ -0,0 +1,227 @@
+ * associatedData is not used by our deterministic encryption, but is used in our unit tests of miscreant encryption.
+ */
+export const encryptBytes = (bytes: Buffer, key: Buffer, associatedData: Uint8Array[] = []): Future<TenantSecurityException, Buffer> => {
+    return Future.tryP(() => miscreant.SIV.importKey(key, "AES-SIV", cryptoProvider))
+        .flatMap((siv) =>
+            Future.tryP(async () => {
+                const encrypted = await siv.seal(bytes, associatedData);
+                return Buffer.from(encrypted);
+            })
+        )
+        .errorMap((e) => new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_ENCRYPT_FAILED, e.message));
+};
+
+export const checkRotationFieldNoOp = (
+    encryptedField: DeterministicEncryptedField,
+    derivedKeys: DerivedKey[] | undefined
+): Future<TenantSecurityException, boolean> =>
+    decomposeField(encryptedField.encryptedField).flatMap((parts) => {
+        const currentKeyId = derivedKeys?.find((key) => key.current)?.tenantSecretId;
+        const previousKeyId = derivedKeys?.find((id) => id.tenantSecretId === parts.tenantSecretId)?.tenantSecretId;
+        if (currentKeyId === undefined || previousKeyId === undefined) {
+            return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_ROTATE_FAILED, "Failed deterministic rotation of field."));
+        } else {
+            return Future.of(previousKeyId === currentKeyId);
+        }
+    });
+
+/**
+ * Decrypt the deterministically encrypted field using the associated tenant secrets contained in `derivedKeys`.
+ * If `reencrypt` is true, return a failure if the field was already encrypted to the current tenant secret.
+ */
+export const decryptField = (
+    encryptedField: DeterministicEncryptedField,
+    derivedKeys: DerivedKey[] | undefined
+): Future<TenantSecurityException, DeterministicPlaintextField> => {
+    const decryptedFuture = decomposeField(encryptedField.encryptedField).flatMap((parts) => {
+        const key = derivedKeys?.find((id) => id.tenantSecretId === parts.tenantSecretId);
+        if (key === undefined) {
+            return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_DECRYPT_FAILED, "Failed deterministic decryption."));
+        } else {
+            return decryptBytes(parts.encryptedBytes, Buffer.from(key.derivedKey, "base64"));
+        }
+    });
+    return decryptedFuture.map((decrypted) => ({
+        plaintextField: decrypted,
+        derivationPath: encryptedField.derivationPath,
+        secretPath: encryptedField.secretPath,
+    }));
+};
+
+/**
+ * Deconstruct the provided encrypted field into its component parts. Separates the tenant secret ID, padding, and encrypted bytes.
+ */
+export const decomposeField = (encryptedBytesWithHeader: Buffer): Future<TenantSecurityException, DeterministicEncryptedFieldParts> => {
+    const tenantSecretIdBytes = encryptedBytesWithHeader.slice(0, 4);
```

Comment:
Do we need to check that the Buffer is at least 6 bytes long? Please write a test for this case.

Reply:
If it's fewer than 6 bytes, then the `padding` will fail to equal `[0, 0]`, so it will error there. I added a test case for it